packages/backend/src/models/JobModel/JobModel.ts

```diff
@@ -0,0 +1,141 @@
+import { NotFoundError } from '../../errors';
+import Logger from '../../logger';
+import Transaction = Knex.Transaction;
+
+type JobModelDependencies = {
+    database: Knex;
+};
+
+export class JobModel {
+    private database: Knex;
+
+    constructor(deps: JobModelDependencies) {
+        this.database = deps.database;
+    }
+
+    async getLastJob(projectUuid: string): Promise<Job | undefined> {
+        const jobs = await this.database(JobsTableName)
+            .where('project_uuid', projectUuid)
+            .orderBy('updated_at', 'desc')
+            .limit(1);
+
+        if (jobs.length === 0) return undefined;
+
+        const job = jobs[0];
+        const steps = await this.getSteps(job.job_uuid);
+
+        return {
+            createdAt: job.created_at,
+            updatedAt: job.updated_at,
+            projectUuid: job.project_uuid,
+            jobUuid: job.job_uuid,
+            jobStatus: job.job_status,
+            steps,
+        };
+    }
+
+    async getJobstatus(jobUuid: string): Promise<Job> {
+        const jobs = await this.database(JobsTableName).where(
+            'job_uuid',
+            jobUuid,
+        );
+
+        if (jobs.length === 0)
+            throw new NotFoundError(
+                `job with jobUuid ${jobUuid} does not exist`,
+            );
+
+        const job = jobs[0];
+        const steps = await this.getSteps(jobUuid);
+        return {
+            createdAt: job.created_at,
+            updatedAt: job.updated_at,
+            projectUuid: job.project_uuid,
+            jobUuid: job.job_uuid,
+            jobStatus: job.job_status,
+            steps,
+        };
+    }
+
+    async getSteps(jobUuid: string): Promise<JobStep[]> {
+        const steps = await this.database(JobStepsTableName)
+            .where('job_uuid', jobUuid)
+            .orderBy('step_id', 'desc');
+
+        return steps.map((step) => ({
+            jobUuid: step.job_uuid,
+            createdAt: step.created_at,
+            updatedAt: step.updated_at,
+            stepStatus: step.step_status,
+            stepType: step.step_type,
+            stepLabel: step.step_label,
+        }));
+    }
+
+    async upsertJobStatus(
+        jobUuid: string,
+        projectUuid: string | undefined,
+        status: JobStatusType,
+    ): Promise<void> {
+        Logger.debug(
+            `Updating job status ${jobUuid} for project ${projectUuid} with status ${status}`,
+        );
+        await this.database(JobsTableName)
+            .insert({
+                project_uuid: projectUuid,
+                job_uuid: jobUuid,
+                job_status: status,
+                updated_at: new Date(),
+            })
+            .onConflict('job_uuid')
+            .merge();
+    }
+
+    async createJobSteps(
+        jobUuid: string,
+        stepTypes: JobStepType[],
+    ): Promise<void> {
+        await this.database.transaction(async (trx) => {
```

Comment:
You can use a single insert with an array of records! That way you don't need to loop at all: https://knexjs.org/#Builder-insert

If there's concern that you want to be able to retrieve them in exactly the right order, SO tells me [postgres doesn't guarantee it'll insert in the same order you specify but it actually does](https://dba.stackexchange.com/questions/95822/does-postgres-preserve-insertion-order-of-records) - we'd have to add an extra column with the order id if we really want to be absolutely certain. But I'm happy to rely on postgres insert order tbh